back to overflowdb traversal, this time with autocompletion working (#338)
ShiftLeftSecurity/joern: docs2/docs/upgrade-guides.mdx

---
id: upgrade-guides
title: Upgrade guides
joern_shared: true
---

## 1.1.1: OverflowDb Traversals
This release introduces a major rearchitecture of the cpg query language (CPGQL). Most changes happened under the hood, however there are a few changes that are user facing. The migration should be straightforward, and in any case we're here to help.

> Background: CPGQL was previously based on tinkerpop gremlin (accessible via `.raw`), and is now based on [OverflowDb Traversal](https://github.com/ShiftLeftSecurity/overflowdb/blob/master/traversal/src/main/scala/overflowdb/traversal/Traversal.scala) instead. The main drivers behind this change are better performance, less complexity and less dependencies. OverflowDb Traversal is a Scala collection with additional graph steps, which means we now inherit many useful steps from the rich Scala standard collection library, where previously we'd need to convert between Traversal and Scala collection. 

Reviewer: ```suggestion
Background: CPGQL was previously based on the [Gremlin graph traversal language](https://tinkerpop.apache.org/gremlin.html) (accessible via `.raw`), and is now based on [OverflowDb Traversal](https://github.com/ShiftLeftSecurity/overflowdb/blob/master/traversal/src/main/scala/overflowdb/traversal/Traversal.scala). The main drivers behind this change are better performance, less complexity and less dependencies. OverflowDb Traversal is a Scala collection with additional graph steps, which means we now inherit many useful steps from the rich Scala standard collection library, where previously we'd need to convert between Traversal and Scala collection.
```

> At the same time, Traversal offers largely the same steps (and semantics) as it's Tinkerpop predecessor. 

Reviewer: ```suggestion
At the same time, Traversal offers largely the same steps and semantics as its TinkerPop predecessor. 
```

> Here are the most important new parts - from our experience the swap of `filter` and `where` accounts for 90% of user-facing changes:

Reviewer: ```suggestion
Here are the most important new parts - from our experience the swap of `filter` and `where` accounts for 90% of user-facing changes:

```

1) `filter(A => Boolean)`: just like any other Scala collection, filter now simply takes a predicate (it used to take a Traversal). 

> 2) `where(trav: Traversal[A] => Traversal[_])`: preserves elements if the provided traversal has at least one result. This is what filter used to be. 

Reviewer: ```suggestion
2. `where(trav: Traversal[A] => Traversal[_])`: preserves elements if the provided traversal has at least one result. This is what filter used to be. *Effectively, `filter` and `where` are swapped*. 

```

A newline after a list item is a Markdown convention that some parsers prefer, I think it would be cool to keep it.
Same with `1)` vs `1.` the latter is more common, and some parsers interpret the previous one differently, it would be nice to stay consistent and use the dot format `1.`

In OverflowDb Traversal this behaviour is more explicit: `.repeat(Traversal)(_.emit)` and `.repeat(Traversal)(_.emitAllButFirst)`.
Some more examples for the migration: 
* `.repeat(x).until(y)` -> `.repeat(x)(_.until(y))`
* `.emit.repeat(x)` -> `.repeat(x)(_.emit)`
* `.repeat(x).emit` -> `.repeat(x)(_.emitAllButFirst)`
* `.emit.repeat(x).times(2)` -> `.repeat(x)(_.emit.times(2))`
5) `repeat` uses depth first search (DFS) by default and can be configured to use breadth first search (BFS) instead. 
Tinkerpop has a long standing [issue](https://github.com/apache/tinkerpop/pull/838) that it claims to use DFS but actually uses BFS, and also cannot be configured to use one or the other. 
6) `.start` step only exists for `Node` and `NewNode`, not for other collections any more. Instead, use the standard scala collection mechanism `.to(Traversal)`:
```
// still works - nothing changed:
val someMethod = cpg.method.head
someMethod.start.parameter //Traversal[MethodParameterIn]

// 
val methodList: cpg.method.l
methodList.to(Traversal) //Traversal[Method]
```
7) `.clone` doesn't exist any more - note that Traversals typically have Iterator semantics, i.e. they are consumed during iteration:

> ```scala

Reviewer: `scala => java`

The code blocks do not support Scala at the moment, Java syntax highlighting is the best we can do. Also, a new line before and after any code block is helpful to the parsers

Author: done
